Add error codes in validate_account_and_subdomain

m-e-l-u-h-a-n opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
issue: #17763 

It defines two new error codes `REALM_DEACTIVATED` and `USER_DEACTIVATED` and uses them in `validate_account_and_subdomain`.

I have not changed error code and message we sent earlier with `JsonableError` , as I was not sure about them also changing them would have required updating many tests just for error message, so I thought of getting reviews over this after creating pr.

**Testing plan:** <!-- How have you tested? -->
Using backend tests.

Fixes: #17763.

timabbott commented on zerver/lib/exceptions.py:
```diff
@@ -271,6 +272,17 @@ def msg_format() -> str:
 
 
+class UserDeactivatedError(JsonableError):
+    code: ErrorCode = ErrorCode.USER_DEACTIVATED
```
I think for these we should probably set `http_status_code=403` -- @mateuszmandera do you agree?

m-e-l-u-h-a-n replied:
I have made `htttp_status_code=403` for now. I'll update it if required.

mateuszmandera replied:
Yeah 403 looks right to me here

timabbott commented on zerver/lib/exceptions.py:
```diff
@@ -271,6 +272,17 @@ def msg_format() -> str:
+    code: ErrorCode = ErrorCode.USER_DEACTIVATED
+
+    def __init__(self) -> None:
+        pass
```
Do we need this `__init__` method?  I haven't read this framework recently, and it seems a bit odd that most of our custom errors need to override it.

m-e-l-u-h-a-n replied:
Yeah I think that's required because if we do not explicitly define it, then when we do `UserDeactivatedError()` python will find no `__init__` method so it will go up in mro hierarchy to call `__init__` method in `JsonableError` call it without `msg` parameter and that will give
```python
TypeError: __init__() missing 1 required positional argument: 'msg'
```
so we need to pass `__init__` in child classes.

timabbott replied:
Ahh, OK, makes sense.

timabbott commented:
Thanks for working on this @m-e-l-u-h-a-n!  I'm also thinking we may want to document these on https://zulip.com/api/rest-error-handling.

@orientor or @MSurfer20 can help you figure out how to edit that page if https://zulip.readthedocs.io/en/latest/documentation/api.html doesn't provide the background needed.

m-e-l-u-h-a-n commented:
@timabbott 
I have added these errors to `/api/rest-error-handling`.
This how `/api/rest-error-handling` looks:
![image](https://user-images.githubusercontent.com/63504956/113020777-8fe6e200-91a0-11eb-842b-002d7b102f4c.png)

This is ready for another review.

timabbott commented on tools/test-api:
```diff
@@ -106,5 +110,16 @@ with test_server_running(
 
+    # Test account deactivated error
+    guest_user.is_active = False
+    guest_user.save()
```
This should be `do_deactivate_user`, I think.

m-e-l-u-h-a-n replied:
used `change_user_is_active` from actions.py along with a comment
```python
# we deactivate user manually because do_deactivate_user removes user session
```
as this is what we do in most of tests in `test_decorators.py` for deactivating users in tests.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -9254,6 +9254,14 @@ paths:
                   - $ref: "#/components/schemas/UserNotAuthorizedError"
+        "403":
+          description: |
+            Forbidden
```
This should comment on the user account being deactivated.

timabbott replied:
Oh, hadn't realized the level this was at -- I think we just was `Forbidden.` to match the "Bad request" styling above? 

(I don't think these end up getting displayed).

m-e-l-u-h-a-n replied:
> (I don't think these end up getting displayed).

Yeah for responses their description is not displayed only example is displayed.

timabbott commented on tools/test-api:
```diff
@@ -121,5 +122,17 @@ with test_server_running(
+    # Test realm deactivated error
+    guest_user.realm.deactivated = True
+    guest_user.is_active = False
+    guest_user.save()
```
Same `do_deactivate_user` comment here -- we try to avoid manually changing objects like this because it can fail to apply other important side effects.

m-e-l-u-h-a-n replied:
Made these changes.

m-e-l-u-h-a-n commented:
I have updated this pr to address comments about using `do_deactivate_user` and similar helpers for deactivating user/realm in `test-api`. I have updated in respective comments about what changes are done.

timabbott commented:
Merged, thanks @m-e-l-u-h-a-n!
